File: app/src/main/java/io/github/romangolovanov/apps/ametro/model/ModelUtil.java
package io.github.romangolovanov.apps.ametro.model;


import android.util.Pair;

import io.github.romangolovanov.apps.ametro.model.entities.MapPoint;
import io.github.romangolovanov.apps.ametro.model.entities.MapRect;
import io.github.romangolovanov.apps.ametro.model.entities.MapScheme;
import io.github.romangolovanov.apps.ametro.model.entities.MapSchemeLine;
import io.github.romangolovanov.apps.ametro.model.entities.MapSchemeStation;

public class ModelUtil {

    public static Pair<MapSchemeLine, MapSchemeStation> findTouchedStation(MapScheme scheme, MapPoint touchPoint) {
        for (MapSchemeLine line : scheme.getLines()) {
            for (MapSchemeStation station : line.getStations()) {
                MapRect rect = station.getLabelPosition();
                MapPoint point = station.getPosition();
                if ((rect != null && rect.contains(touchPoint)) ||
                        (point != null && point.distance(touchPoint) <= scheme.getStationsDiameter())) {
                    return new Pair<>(line, station);
                }
            }
        }
        return null;
    }

    public static Pair<MapSchemeLine, MapSchemeStation> findStationByUid(MapScheme scheme, long uid) {
        for (MapSchemeLine line : scheme.getLines()) {
            for (MapSchemeStation station : line.getStations()) {
                if (station.getUid() == uid) {
                    return new Pair<>(line, station);
                }
            }
        }
        return null;
    }

}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/model/ModelUtil.kt
package io.github.romangolovanov.apps.ametro.model

import io.github.romangolovanov.apps.ametro.model.entities.MapPoint
import io.github.romangolovanov.apps.ametro.model.entities.MapSchemeLine
import io.github.romangolovanov.apps.ametro.model.entities.MapScheme
import io.github.romangolovanov.apps.ametro.model.entities.MapSchemeStation

object ModelUtil {

    fun findTouchedStation(scheme: MapScheme, touchPoint: MapPoint): Pair<MapSchemeLine, MapSchemeStation>? {
        return scheme.lines.firstNotNullOfOrNull { line ->
            line.stations.firstOrNull { station ->
                val rect = station.labelPosition
                val point = station.position
                (rect != null && rect.contains(touchPoint)) ||
                    (point != null && point.distance(touchPoint) <= scheme.stationsDiameter)
            }?.let { line to it }
        }
    }

    fun findStationByUid(scheme: MapScheme, uid: Long): Pair<MapSchemeLine, MapSchemeStation>? {
        return scheme.lines.firstNotNullOfOrNull { line ->
            line.stations.firstOrNull { it.uid.toLong() == uid }?.let { line to it }
        }
    }
}
